File: apps/sdkwork-birdcoder-flutter-mobile/packages/sdkwork_birdcoder_flutter_mobile_core/lib/src/bootstrap/auth_oauth_deep_link.dart
const birdCoderMobileOAuthScheme = 'birdcoder';
const birdCoderOAuthCallbackAuthority = 'auth';
const birdCoderOAuthCallbackPath = '/oauth/callback';

String buildBirdCoderOAuthCallbackReturnUrl({String? provider}) {
  final uri = Uri(
    scheme: birdCoderMobileOAuthScheme,
    host: birdCoderOAuthCallbackAuthority,
    path: birdCoderOAuthCallbackPath,
  );
  final normalizedProvider = provider?.trim();
  if (normalizedProvider == null || normalizedProvider.isEmpty) {
    return uri.toString();
  }
  return uri.replace(queryParameters: {'provider': normalizedProvider}).toString();
}

String? readBirdCoderOAuthProviderFromCallbackPath(String path) {
  final match = RegExp(r'/oauth/callback/([^/?#]+)').firstMatch(path);
  final provider = match?.group(1)?.trim();
  if (provider == null || provider.isEmpty) {
    return null;
  }
  return provider;
}

File: apps/sdkwork-birdcoder-h5/packages/sdkwork-birdcoder-h5-core/src/bootstrap/authOAuthDeepLink.ts
import { BIRDCODER_AUTH_BASE_PATH } from '@sdkwork/birdcoder-pc-auth';

export const BIRDCODER_H5_OAUTH_SCHEME = 'birdcoderh5';
export const BIRDCODER_H5_OAUTH_CALLBACK_AUTHORITY = 'auth';
export const BIRDCODER_H5_OAUTH_CALLBACK_PATH = '/oauth/callback';

export function buildBirdCoderH5OAuthCallbackReturnUrl(provider?: string): string {
  const url = new URL(
    `${BIRDCODER_H5_OAUTH_SCHEME}://${BIRDCODER_H5_OAUTH_CALLBACK_AUTHORITY}${BIRDCODER_H5_OAUTH_CALLBACK_PATH}`,
  );
  const normalizedProvider = provider?.trim();
  if (normalizedProvider) {
    url.searchParams.set('provider', normalizedProvider);
  }
  return url.toString();
}

function joinAuthPath(path: string, search: string): string {
  const normalizedPath = path.replace(/\/+$/, '') || '/';
  return search ? `${normalizedPath}${search}` : normalizedPath;
}

function isBirdCoderAuthPath(path: string): boolean {
  const normalized = path.replace(/\/+$/, '') || '/';
  return normalized === BIRDCODER_AUTH_BASE_PATH
    || normalized.startsWith(`${BIRDCODER_AUTH_BASE_PATH}/`);
}

export function normalizeBirdCoderH5AuthDeepLinkPath(rawUrl: string): string | null {
  try {
    const uri = new URL(rawUrl);
    const search = uri.search || '';
    const candidates = [
      joinAuthPath(uri.pathname, search),
      uri.host ? joinAuthPath(`/${uri.host}${uri.pathname}`, search) : '',
      uri.host && !uri.pathname ? joinAuthPath(`/${uri.host}`, search) : '',
    ];

    for (const candidate of candidates) {
      if (isBirdCoderAuthPath(candidate)) {
        return candidate;
      }
    }
  } catch {
    return null;
  }

  return null;
}
